### PR TITLE
Handle unrecognized input option; simplified readparamters

### PR DESCRIPTION
**Pull request by henry2004y**
I propose two things in this PR:

1. Abort the code when any unrecognized input option from cfg file is detected, so that we can avoid typos and misplacement of  options (e.g. `nVelocitySamples` under multiple commands). The current dev branch allows reading unrecognized options and the user is not aware whether he or she has made a mistake in the cfg file. For example
```yaml
[gridbuilder]
x_length = 20
# comments are not affected
y_lenght = 20 # typo, ignored in current dev, caught in this PR
z_length = 1 
```
This PR will display error message like
```
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::program_options::unknown_option> >'
  what():  unrecognised option 'gridbuilder.y_lenght'
```

2.  A simplified and extended implementation of `Readparameters`. The current dev branch applies function overloading over `add` and `get` methods, which requires the programmer to explicitly list all the possible input types. The new implementation uses templates. Besides, error handling is done within the `add` and `get` methods, so we no longer need to return a `bool` type:
```c++
// current dev pattern
int myRank;
MPI_Comm_rank(MPI_COMM_WORLD,&myRank);
if(!Readparameter::get("Template.param", this->param)) {
if(myRank == MASTER_RANK) cerr << __FILE__ << ":" << __LINE__ << " ERROR: This option has not been added!" << endl;
   exit(1);
}

// proposed new approach, error handling is done within the method
Readparameter::get("Template.param", this->param);
```
Since we can specify the error message specifically for the input string argument, this will not create extra burden for debugging.

I have only run one test, `Flowthrough_x_inflow_y_outflow`.

### CODE REVIEW

**Comment by ykempf** [comment]
I think we all agree that this is a very nice contribution. I suggest that you run all the testpackage cases to clean up their cfgs, I'd be surprised if they didn't have stale lines.

What do @ursg and @markusbattarbee say, should we use the occasion to clean up the projects' cfgs, or do we leave that for a later stage and/or delegate the task?

File: parameters.cpp
Comment (ykempf) [inline parameters.cpp:151]: Let's hope this doesn't introduce int vs uint warnings.

File: readparameters.cpp
Comment (ykempf) [inline readparameters.cpp:344]: As @markusbattarbee noted, it owuld be good to add this new line here.

File: readparameters.h
Comment (ykempf) [inline readparameters.h:190]: Also here, new line.

File: vlasiator.cpp
Comment (ykempf) [inline vlasiator.cpp:888]: "system sysBoundaries" is a bit of a repeat now, either "system boundaries" or "sysBoundaries" I'd say.

File: vlasiator.cpp
Comment (ykempf) [inline vlasiator.cpp:960]: Ditto

**Comment by markusbattarbee** [comment]
Yep, agreed, this is very nice. Indeed, running the standard testpackage to check that all inputs are still parsed correctly is what's needed.

I don't think we need to keep this PR waiting for cleaning up of less-used old configs - especially as it isn't clear what needs to be done to them. Output variable names have already been updated, and I guess they should be multipop-compatible, unless in the unsupported subdirectory?

One minor point is that in the already-accepted-but-not-merged PR #543 I moved the shortPencils flag under the AMR group, so whichever PR gets merged last needs to be updated to reflect that change.

**Review by ykempf — COMMENTED** [review]
You'll end up with more code deletions than additions on the balance. But it'll look much better, and the template approach is so much more elegant. We just didn't have/use templates then.

File: readparameters.h
Comment (ursg) [inline readparameters.h:41]: Hmm, in principle, it is bad style to declare a namespace shorthand like this in a header file.
After all, now every c++ file that includes it will also have the PO name resolved to boost::program_options.

I see that this is unavoidable for a template implementation (unless we want to spell out boost::program_options in every point it is used, which is not pretty, either), but I'm worried that "PO" might be an identifier somebody accidentally uses at one point, and is then confused by the error message.

File: readparameters.h
Comment (ursg) [inline readparameters.h:66]: I'm wondering (without having tested): Does this need a separate specialization for ```const char *```? In the current implementation, there's the issue that ```const char *``` is treated as a numeric type, since it can be typeconverted to one. Hence, their string parameter values are typeconverted to a number, which ends up being 0, so the string becomes a nullpointer at runtime.

The "proper" way to use string parameters is ```RP::get("name", "description", std::string("default"));``` at the moment. I wondr if a simple extra specialization would fix this.

File: readparameters.h
Comment (ursg) [inline readparameters.h:116]: Hmm, how does the return type (void) play with our existing use pattern here? In all old code, ```RP::get()``` is called like
```
if(!RP::get("name",target)) {
   std::cerr << "Some error message" << std::endl;
}
```
So if this becomes void, that line will throw an error.

Since you changed all calls to this in all projects anyway (awesome work! I think it makes the code a lot better), but I'm wondering how this'll affect any as-yet-unmerged pull requests. Three options:
 - If we leave this to be a bool, they will probably simply work. But their code will probably never be cleaned up, because why would it?
 - If we leave it as void, they will throw errors, and whoever works on them gets to fix their code once they merge.
 - The ideal case would be to set it to bool and make the compiler emit a warning if it is actually used. Something like the opposite of [the nodiscard attribute](https://en.cppreference.com/w/cpp/language/attributes/nodiscard) or the opposite of [GCC's warn_unused_result](https://gcc.gnu.org/onlinedocs/gcc-4.1.1/gcc/Function-Attributes.html) attribute. Except these don't seem to exist. Ho-hum.

File: vlasiator.cpp
Comment (ursg) [inline vlasiator.cpp:482]: Uh-oh, since readparameters is an auto variable, this explicit call of th destructor is a bit dangerous.
Especially since it calls unprotected ```delete``` on some pointers in there. Won't this cause a double free, when th readparameters object is eventually deconstructed at the end of main?
Not that it would really matter if it crashes at the very end, but I would suggest putting a nullpointer check around the deletes in the destructor.

**Review by ursg — APPROVED** [review]
Awesome! I really like how this makes the project and boundary code more readable, since it gets rid of lots of boilerplate.
Some commnts below, but they are not showstoppers. We could just merge this as-is, and fix it in dev if any issues with other branches still arise.

File: parameters.cpp
Comment (ursg) [reply parameters.cpp:151]: I actually believe it fixes them, instead of introducing them.

Comment (henry2004y) [reply parameters.cpp:151]: There are several places where we have loops over AMR widths and levels with local variable of type int. I thought `auto` would do the work, but in fact I had to declare them as `uint` to avoid the warnings. Some of them have already been modified in this commit. I will check again.

File: readparameters.h
Comment (henry2004y) [reply readparameters.h:190]: Will do

File: parameters.cpp
Comment (ykempf) [reply parameters.cpp:151]: Perfect then. Anyway this is used rarely and I don't think it will cause issues as this won't overflow, nor will it be set to negative values.

File: vlasiator.cpp
Comment (henry2004y) [reply vlasiator.cpp:888]: I think I messed up with my other branch where `sysboundary` are renamed to `boundary`. I will revert the changes here.

File: parameters.cpp
Comment (henry2004y) [reply parameters.cpp:625]: Will do

File: vlasiator.cpp
Comment (ykempf) [reply vlasiator.cpp:888]: ok

File: readparameters.h
Comment (henry2004y) [reply readparameters.h:41]: Indeed. I will remove using namespaces in headers.

File: readparameters.h
Comment (henry2004y) [reply readparameters.h:116]: I notice that in some of the projects, the `get` function in current dev is called as
```c++
RP::get("name", target)
```
which does not follow the design pattern but it works as long as no error in reading occurs.

I don't know which is a proper way to handle this backward compatibility like problem. Leave for you to decide.

File: vlasiator.cpp
Comment (henry2004y) [reply vlasiator.cpp:482]: This is inherited from the original struct of `Readparameters`, but I agree it is dangerous. I kept it here as it was because maybe freeing the memory manually here would save some memory in the actual time marching loops. 

One thing I noticed in this destructor
```c++
Readparameters::~Readparameters() {
   delete descriptions;
   delete variables;
   descriptions = NULL;
   variables = NULL;
}
```
was that without setting these two pointers to `NULL`, the program will crash in the end. I am not sure why. Will there be an enforced auto free at the end of `main()`? What should I do to make it better?

Comment (ursg) [reply vlasiator.cpp:482]: The readparameters object itself will fall out of scope at the end of main, so the destructor will run again in that case. If you don't set them to NULL, that will be a certain double free. This probably causes the crash.

We could either guard the deletes:
```c++
if(descriptions != nullptr) {
   delete descriptions;
}
```
or we decide that we don't care about this little bit of wasted memory in a simulation that's multiple gigabytes heavy anyway.

File: readparameters.h
Comment (henry2004y) [reply readparameters.h:65]: I haven't seen an example for parsing character type. Perhaps it would be better to treat everything as string first and then do the type conversion. If we want to test this then perhaps we should add a test for this edge case. For now I will ignore this potential issue.

**Comment by henry2004y** [comment]
I think it's ready. Tested on my local ubuntu laptop with testpackage tests. No abort found in existing tests' cfg files.

**Comment by markusbattarbee** [comment]
> I think it's ready. Tested on my local ubuntu laptop with testpackage tests. No abort found in existing tests' cfg files.

Did the test outputs also match the previous version? (i.e. did all the accepted parameters get parsed correctly?)

**Comment by henry2004y** [comment]
> > I think it's ready. Tested on my local ubuntu laptop with testpackage tests. No abort found in existing tests' cfg files.
> 
> Did the test outputs also match the previous version? (i.e. did all the accepted parameters get parsed correctly?)

Not all of them. For example, `transtest_amr` has differences within `1e-8` with single precision VDFs. I am not 100% sure if it works for all after all the changes with the current testpackage settings, because some of them take too long to run on a laptop. So I just decrease the timesteps and compare with that.

**Comment by ykempf** [comment]
Well single precision doesn't help... It should be run with testpackage flags. But reducing the timesteps is OK I think.

On 15 July 2021 08:53:02 EEST, Hongyang Zhou ***@***.***> wrote:
>> > I think it's ready. Tested on my local ubuntu laptop with
>testpackage tests. No abort found in existing tests' cfg files.
>> 
>> Did the test outputs also match the previous version? (i.e. did all
>the accepted parameters get parsed correctly?)
>
>Not all of them. For example, `transtest_amr` has differences within
>`1e-8` with single precision VDFs. I am not 100% sure if it works for
>all after all the changes with the current testpackage settings,
>because some of them take too long to run on a laptop. So I just
>decrease the timesteps and compare with that.
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/fmihpc/vlasiator/pull/561#issuecomment-880416152

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by henry2004y** [comment]
Tested with testpackage settings, double precision VDFs. Exactly identical output files, except `Magnetosphere_3D_small`, where the saved variables are exactly the same but the system diff showed that the two files are different.

**Comment by ykempf** [comment]
Great then! :)

File: sysboundary/sysboundary.cpp
Comment (markusbattarbee) [inline sysboundary/sysboundary.cpp:85]: I only now (when trying to bring another branch up-to-date and finding conflicts) see that here the default value of periodicity has changed. 

Although it can make sense (e.g. if no other boundary has been set, then assume periodicity) it should have been documented in the PR. If some other boundary type is set, does that now automatically unset periodicity, or can there be a clash?

**Pull request merged**
